torchvision/prototype/models/resnet.py

```diff
@@ -231,11 +242,28 @@ class WideResNet101_2Weights(Weights):
     )
 
 
+# Maybe we should rename kwargs into model_kwargs to be more explicit? This is not a BC-breaking change I believe.
 def resnet18(weights: Optional[ResNet18Weights] = None, progress: bool = True, **kwargs: Any) -> ResNet:
+    # Considering we're showing this warning literally everywhere, it might be worth writing a helper for it
+    # This would make sure the warning message is consistent, avoid code duplication, etc.
     if "pretrained" in kwargs:
-        warnings.warn("The argument pretrained is deprecated, please use weights instead.")
+        # Nit: "argument" should probably be "parameter", since the term
+        # "argument" refers to the *value* of the parameter at call time.
+        warnings.warn("The 'pretrained' parameter is deprecated, please use the 'weights' parameter instead.")
         weights = ResNet18Weights.ImageNet1K_RefV1 if kwargs.pop("pretrained") else None
 
+    # To preserve full BC with the current resnet18, weights coudl also accept
+    # a boolean and this would issue deprecation warning as well. This is for
+    # cases where resnet18 is called with resnet18(True) or resnet18(False)
+    # something like this:
+    if isinstance(weights, bool):
+        warnings.warn("The 'pretrained' parameter is deprecated, please use the 'weights' parameter instead.")
+        weights = ResNet18Weights.ImageNet1K_RefV1 if weights else None
+
+    # I remember we discussed this before: before releasing this new API, we
+    # mght want to try to allow users to pass simple stuff like e.g.
+    # resnet18(weights='pretrained') or resnet18(weights='latest')
```

Comment:
Concerning `weights='latest'`, yes we will eventually support it. We had an [early version](https://github.com/datumbox/dapi-model-versioning/blob/main/dapi_lib/models/_api.py#L157-L159) of this on an old prototype. This will be handled once we have the registration mechanism in place. I'm delaying including this to see if we can achieve some cross-domain agreement. Could you clarify if you think this is a prerequisite for releasing or if this can be implemented later? I assume your concern is about TorchHub compatibilities.

Could you clarify if you literally meant that we should support `weights='pretrained'`? If yes, what does this do? Note that currently we support `weights="ImageNet1K_RefV1"`. Would that be good enough?

Reply:
> Could you clarify if you think this is a prerequisite for releasing or if this can be implemented later?

I wouldn't make this a requirement for merging the prototype into `main`, but I think we should support this for the release. Currently users can get pretrained weights with a super simple api, just passing `pretrained=True`. Now we offer a more powerful API but we should still try to support these very simple workflows IMO. If we're telling them "don't use this, use that instead", using "that" should be as painless as possible. Ideally we would just tell them in the deprecation warning something like

> The 'pretrained' parameter is deprecated, please use the 'weights' parameter instead. 'pretrained=True' is now equivalent to 'weights="latest"' (or 'pretrained' or 'default' or whatever we want to decide)





> Could you clarify if you literally meant that we should support weights='pretrained'?

Maybe 'pretrained' isn't a good name. But I think that ideally we would allow users to get both a) the latest weights and b) the "default" weights (which would never change): the latest weights might change across versions and for some users the changes might not be desirable.

Reply:
> But I think that ideally we would allow users to get both a) the latest weights and b) the "default" weights (which would never change)

No concerns for a), it's going to be supported soon.

Concerning b), it was a conscious choice to set `weights=None` for default instead of `weights=SomeValue` in the method signature. The idea was that the default weights might become obsolete and then we would be stuck keeping them in the signature due to the BC considerations of TorchVision. By forcing the users to explicitly define the weights we avoid that. I think asking the user to pass as a string the weights that they want, is a reasonable compromise between easy of use and keeping the API fresh?

Reply:
I'm not suggesting to change the default of `weights`, it's fine to leave it as None.

What I'm saying is that we might want to support an equivalent behaviour as the current `pretrained=True`, i.e. allow the user to pass `weights=<something>` (where `<something>` is to be determined) that will always return the same pre-trained weights.

Reply:
> I'm not suggesting to change the default of weights, it's fine to leave it as None.

I understood that. Sorry I should have been more explicit on how I made the link between `weights='pretrained'` and the default values of `weights` in the signature.

The reason I linked the two is because IMO having a non-None value as default has exactly the same effect as supporting something like `weights='pretrained'`. They anchor us to a specific old and potentially superseded set of weights. That's why I'm in favour of supporting still a `weights=<something>` but that something shouldn't be constant across all models but rather explicitly define which set of weights to load. This is currently already supported because the weights can be a string with the name of the Enum.

In other words, I'm hoping to avoid introducing anything on the API that will anchor the user to legacy weights. I'm happy to support a `weights='latest'` thought, for those who are happy to always access the best available weights for a specific model architecture.

Reply:
Thanks for clarifying @datumbox , that's fair.

My main concern here is to be as nice as possible to users and help them transitioning from the old to the new API. Without introducing a `weights='pretrained'` option, perhaps it would be enough to expand the deprecation message a bit to e.g.:

> The 'pretrained' parameter is deprecated, please use the 'weights' parameter instead. 'pretrained=True' is now equivalent to 'weights="TheWeightsNameWithIsSpecificToEachModelBuilder"'; you can also use 'pretrained="latest"' to get the most up-to-date weights for this model. For more details see \<link to doc note where we explain the changes introduced with this API\>.

(instead of suggesting a string like "TheWeightsNameWithIsSpecificToEachModelBuilder" we can also suggest the enum field, not strong opinion).

Reply:
> I wouldn't make this a requirement for merging the prototype into main, but I think we should support this for the release

BTW, if we go with a fully detailed deprecation message as above, I would say that it's OK to release without the weights='lastest' option (although still preferable)

Reply:
I agree with the approach of the fully detailed deprecation message. I think it's a reasonable compromise.

For the `weights='latest'`, I don't really have concerns and I would actually like to give the option to users to access the best/greatest available weights. Not all people in the industry are interested in BC guarantees on the weights side. Sometimes you just want to get the best one and continue your work on transfer learning.

Reply:
@NicolasHug We could do a "poor man's implementation" of the `latest` feature by literally setting the value to the selected weight within the builder. This is an implementation detail invisible to the API and can be replaced with a fancy technique once we have the registration mechanism. I'm tempted to implement this.

Reply:
Yes, that sounds reasonable

Reply:
@NicolasHug @fmassa I'm just throwing this out there: 

An alternative way to implement this `latest weights` feature could be to maintain `pretrained=True` as a shortcut to get always the most updated weights. This USED to be the case since all models had only a single weight. There were even cases where old weights moved to `pretrained="legacy"` (see [here](https://github.com/pytorch/vision/blob/main/torchvision/models/detection/keypoint_rcnn.py#L379-L380)). So in a sense one could argue that the default behaviour of `pretrained=True` is currently to return the best available weights, hence it's not BC-breaking. Note that this approach is followed by other libraries where they overwrite directly the weights in-place.

Doing this will get us automatic adoption to the best/greatest weights while at the same time it will allow users who want to always access specific weights to be able to do so. Thoughts?

Reply:
The above comment will be discussed separately.

For the original request discussed here, I implemented a basic mechanism that allows us to define `weights="default"` at PR #4948 commit 8aa1c233dbf71790235840e8448a34ad43217d5c. I preferred `default` instead of `latest` because it describes better what it does and gives us the freedom to update the default weights for reasons other than recency. This is particularly useful if we have multiple weights on different datasets (for example Voc vs COCO) but we can have only 1 default.